gpx.cc

```diff
@@ -421,34 +424,44 @@ xml_parse_time(const QString& dateTimeString)
         }
       }
     }
-  }
-
-  double fsec = 0;
-  char* pointstr = strchr(timestr, '.');
-  if (pointstr) {
-    sscanf(pointstr, "%le", &fsec);
-#if 0
-    /* Round to avoid FP jitter */
-    if (microsecs) {
-      *microsecs = .5 + (fsec * 1000000.0) ;
+#else
+    // static const QRegularExpression re(R"([+-])(\d+):(\d+)");
+    static const QRegularExpression re("([+-])\(\\d+):(\\d+)");
+    QRegularExpressionMatch match = re.match(dateTimeString);
+    assert(re.isValid());
+    if (match.hasMatch()) {
+      off_sign = match.captured(1) == '-' ? -1 : 1;
+      off_hr = match.captured(2).toInt();
+      off_min = match.captured(3).toInt();
     }
 #endif
-    *pointstr = '\0';
   }
-
-  int year = 0, mon = 0, mday = 0, hour = 0, min = 0, sec = 0;
+if (off_hr) qDebug() << dateTimeString << off_hr << off_min;
   gpsbabel::DateTime dt;
-  int res = sscanf(timestr, "%d-%d-%dT%d:%d:%d", &year, &mon, &mday, &hour,
+#if 0
+  int year = 0, mon = 0, mday = 0, hour = 0, min = 0;
+  double sec;
+  int res = sscanf(timestr, "%d-%d-%dT%d:%d:%lf", &year, &mon, &mday, &hour,
                    &min, &sec);
   if (res > 0) {
+#else
+  static const QRegularExpression
+    re("(\\d+)-(\\d+)-(\\d+)T(\\d+):(\\d+):([0-9.]+)");
+  QRegularExpressionMatch match = re.match(dateTimeString);
+  assert(re.isValid());
+  if (match.hasMatch()) {
+    int year = match.captured(1).toInt();
+    int mon = match.captured(2).toInt();
+    int mday = match.captured(3).toInt();
+    int hour = match.captured(4).toInt();
+    int min = match.captured(5).toInt();
+    double sec = match.captured(6).toDouble();
+#endif
     QDate date(year, mon, mday);
-    QTime time(hour, min, sec);
-    dt = QDateTime(date, time, Qt::UTC);
+    // The ctor for time accepts int secs, so we have to break out the msecs.
+    QTime time(hour, min, sec, ((sec - (int)sec + .0005) * 1000));
 
-    // Fractional part of time.
-    if (fsec) {
-      dt = dt.addMSecs(lround(fsec * 1000));
-    }
+    dt = QDateTime(date, time, Qt::UTC);
 
     // Any offsets that were stuck at the end.
     dt = dt.addSecs(-off_sign * off_hr * 3600 - off_sign * off_min * 60);
```

Comment:
I like the replacement of sscanf with QRegularExpression, and elimination of cstdio, cstring.

I had WIP, including a bunch of issues that appear with Qt 6.5.0, with an idea regarding offsets.  Reviewing your ideas and incorporating mine got messy, a modified version to compare is attached.  To use this also change the declaration in defs.h adding a default to match current code.
gpsbabel::DateTime xml_parse_time(const QString& dateTimeString, bool local_is_utc = true);

I modified the regular expressions based on https://www.w3.org/TR/xmlschema-2/#dateTime.  I didn't fret over negative years, or years beyond 9999.

I do worry a bit about the offset regular expression matching something it shouldn't.  I haven't thought of a way it could mismatch something besides a timezone from https://www.w3.org/TR/xmlschema-2/#isoformats

qRound and qRound64 let you round a double to a int or qint64.

It also got messy with the conditional code and keeping them both running, I did the lazy thing and chopped it.

I haven't tried the enhanced kml tests to see if they work with these changes.

I attached the modified gpx.cc, or a patch for both gpx.cc and defs.h:
[gpx.cc.gz](https://github.com/GPSBabel/gpsbabel/files/11472826/gpx.cc.gz)
[xsddatetime.patch](https://github.com/GPSBabel/gpsbabel/files/11472837/xsddatetime.patch)